Reintroduce email unicity validation

bibips opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project!

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/#pull-requests

For type and category see:
https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#type--category
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | The validation of the unicity of the email address is reinstated. This validation was in place in [version 8.2](https://github.com/PrestaShop/PrestaShop/blob/8.2.x/classes/form/CustomerForm.php#L129) but is not present in the [version 9.0](https://github.com/PrestaShop/PrestaShop/blob/develop/classes/form/CustomerForm.php#L124)
| Type?             | bug fix
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | 1. Go to customer identity page<br>2. Change the email and set an email already used by another customer<br>3. An error message is displayed
| UI Tests          | https://github.com/bibips/ga.tests.ui.pr/actions/runs/17938704488
| Fixed issue or discussion?     | Fixes #39587

bibips committed:
Fix(customer): reintroduce email unicity validation

Codencode commented:
@Hlavtox @kpodemski Let's wait for others' opinions to understand how to proceed, or should we approve the PR?

Hlavtox commented on classes/form/CustomerForm.php:
```diff
@@ -123,6 +123,17 @@ public function getCustomer()
 
     public function validate()
     {
+        if (!$this->context->customer->is_guest) {
+            $emailField = $this->getField('email');
+            if (Customer::customerExists($emailField->getValue())) {
+                $emailField->addError($this->translator->trans(
+                    'The email is already used, please choose another one or sign in',
+                    [],
+                    'Shop.Notifications.Error'
+                ));
+            }
+        }
```
```suggestion
        // If email is being changed, check if it's free to use and doesn't belong to another customer
        if (!$this->context->customer->is_guest) {
            $emailField = $this->getField('email');
            if (Customer::customerExists($emailField->getValue())) {
                $emailField->addError($this->translator->trans(
                    'The email is already used, please choose another one',
                    [],
                    'Shop.Notifications.Error'
                ));
            }
        }
```

Hlavtox reviewed (commented):
Minor suggestion, otherwise I think it's good. Getting customer through context is legit, the class already does this.

SiraDIOP reviewed (commented):
Hello @bibips 
Thank you for the pr. Could you restart the automated tests please?

Thanks

Hlavtox committed:
Improve email validation in CustomerForm

Enhance email validation to check if the email belongs to another customer when changing it.

jolelievre commented on classes/form/CustomerForm.php:
```diff
@@ -123,6 +123,20 @@ public function getCustomer()
     public function validate()
     {
+        // If email is being changed, check if it's free to use and doesn't belong to another customer
+        if (!$this->context->customer->is_guest) {
```
Be careful if you put this check here it means it will only be used in the FO form, so not in BO, not in webservice and not in Admin API

Maybe this check should be placed in a protected method in `Customer::validateEmailUnicity` instead, and this method should be called in add/update/save methods automatically

Or alternatively we could create a new dedicated validation rule for legacy fields and configure it on this email field but I'm not sure if it's possible this way

Codencode replied:
Hello @jolelievre, if I'm not mistaken and if I understood correctly, these checks should have already been included by this separate PR: https://github.com/PrestaShop/PrestaShop/pull/39591

Hlavtox replied:
@jolelievre It's already implemented in other places, we didn't realize / test this use case only.

jolelievre replied:
Hmm, but I don't understand if the tests are already done on the Customer level why is it needed to do them on the corm level as well? The Customer should trigger an invalid data no?

Hlavtox replied:
@jolelievre It (the PR for Customer class) would also catch it, as a second safeguard, but it would just throw a generic error on the top of the page.

This PR (we are currently discussing) will throw a precise error under the email field.

jolelievre replied:
Ok got it, thanks

AureRita reviewed (approved):
Hi @bibips 

Thank you for your PR, I tested it and it seems to works as you can see : 

https://github.com/user-attachments/assets/36fcd4df-421e-4f1e-8018-31a37a9d21ee

link of automated test : https://github.com/AureRita/testing_pr/actions/runs/17947279191

Because the PR seems to works as expected, It's QA :heavy_check_mark: 

Thank you
